bitutils/bitutils.go

```diff
@@ -0,0 +1,96 @@
+package bitutils
+
+import (
+	"encoding/binary"
+	"fmt"
+)
+
+// ParseByte4 parses 4 bits of data from the data array, starting at the given index
+func ParseByte4(data []byte, bitStartIndex uint) (byte, error) {
+	startByte := bitStartIndex / 8
+	bitStartOffset := bitStartIndex % 8
+	if bitStartOffset < 4 {
+		if uint(len(data)) < (startByte + 1) {
+			return 0, fmt.Errorf("ParseByte4 expected 4 bits to start at bit %d, but the consent string was only %d bytes long", bitStartIndex, len(data))
+		}
+		return (data[startByte] & (0xf0 >> bitStartOffset)) >> (4 - bitStartOffset), nil
+	}
+	if uint(len(data)) < (startByte+2) && bitStartOffset > 3 {
+		return 0, fmt.Errorf("ParseByte4 expected 4 bitst to start at bit %d, but the consent string was only %d bytes long", bitStartIndex, len(data))
+	}
+
+	leftBits := (data[startByte] & (0xf0 >> bitStartOffset)) << (bitStartOffset - 4)
+	bitsConsumed := 7 - bitStartOffset
+	overflow := 4 - bitsConsumed
+	rightBits := (data[startByte+1] & (0xf0 << (4 - overflow))) >> (8 - overflow)
+	return leftBits | rightBits, nil
+}
```

Comment:
I'm showing my work in validating this function. Not all of these test cases are accounted for in the unit tests and it looks like there are failures.

Assumptions:
- bitStartIndex is 0-bound.

Case 1: bitStartIndex perfectly aligns with a byte / first nibble [1 byte needed]
```
data: 00010000 (0x10). (desired output: first 4 bits = 1)
bitStartIndex: 0
startByte: 0
bitStartOffset: 0
bitStartOffset < 4: true
uint(len(data)) < (startByte + 1)
= 1 < 1: no error
(data[startByte] & (0xf0 >> bitStartOffset)) >> (4 - bitStartOffset), nil
= (0x10 & 0xf0) >> 4
= 1 (correct)
```

Case 2: bitStartIndex perfectly aligns with an odd nibble [1 byte needed]
```
data: 10010010 (0x92). (desired output: last 4 bits = 2)
bitStartIndex: 4
startByte: 0
bitStartOffset: 4
bitStartOffset < 4
= 4 < 4: false
uint(len(data)) < (startByte+2) && bitStartOffset > 3
= 1 < 2 && 4 > 3
= true && true: error (incorrect)
```

Case 3: bitStartIndex doesn't perfectly align with a nibble, start in first 4 bits [1 byte needed]
```
data: 10011001 (0x99) (desired output: bits 1 through 5 = 3)
bitStartIndex: 1
startByte: 0
bitStartOffset: 1
bitStartOffset < 4 
= 1 < 4: true
uint(len(data)) < (startByte + 1)
= 1 < 1: false
(data[startByte] & (0xf0 >> bitStartOffset)) >> (4 - bitStartOffset)
= (0x99 & (0xF0 >> 1) >> 3
= (0x99 & 0x78) >> 3
= 0x18 >> 3 = 3 (correct)
```

Case 4: bitStartIndex doesn't perfectly align with a nibble, start in last 4 bits [2 bytes needed]
```
data: 00000001 (0x01), 11100000 (0xE0) (desired output bits 7 through 11 = 0XF)
startByte: 0
bitStartOffset: 7
bitStartOffset < 4
= 7 < 4: false
uint(len(data)) < (startByte+2) && bitStartOffset > 3
= 2 < 2 && 7 > 3
= false && true 
= false
leftBits := (data[startByte] & (0xf0 >> bitStartOffset)) << (bitStartOffset - 4)
= (0x01 & (0xF0 >> 7)) << 3
= (0x01 & 0x01) << 3
= 0x08
bitsConsumed := 7 - bitStartOffset
= 7 - 7
= 0
overflow := 4 - bitsConsumed
= 4 - 0
= 4
rightBits := (data[startByte+1] & (0xf0 << (4 - overflow))) >> (8 - overflow)
= (0xE0 & (0xF0 << (4 - 4)) >> (8 - 4)
= (0xE0 & 0xF0)  >> 4
=  0xE0 >> 4
=  0x0E
leftBits | rightBits
= 0x08 | 0x0E
= 0xE (incorrect)
```

Reply:
This was a bit time consuming. I'll leave this here for now and suggest improved test coverage.